synapse/handlers/pagination.py

```diff
@@ -132,6 +136,82 @@ def __init__(self, hs: "HomeServer"):
                     job.longest_max_lifetime,
                 )
 
+        if self._is_master:
+            self.clock.looping_call(
+                run_as_background_process,
+                PaginationHandler.PURGE_ROOMS_PERIOD,
+                "purge_rooms",
+                self.purge_rooms,
+            )
+
+    async def purge_rooms(self) -> None:
+        """This takes care of restoring unfinished purge/shutdown rooms from the DB.
+        It also takes care to launch scheduled ones, like rooms that has been fully
+        forgotten.
+
+        It should be run regularly.
+        """
+        rooms_to_purge = await self.store.get_rooms_to_purge()
+        for r in rooms_to_purge:
+            room_id = r["room_id"]
+            delete_id = r["delete_id"]
+            status = r["status"]
+            timestamp = r["timestamp"]
+
+            if (
+                status == DeleteStatus.STATUS_COMPLETE
+                or status == DeleteStatus.STATUS_FAILED
+            ):
+                # remove the delete from the list 24 hours after it completes or fails
```

Comment:
It makes sense to clean-up after the delete completes but it's less clear why when it fails. We should add some context why when it fails, perhaps because we have no way to retry and is bound to fail again and people should start a new shutdown/purge action.

Reply:
We don't do retry after failed at all. For now it's left at the user initiative.
I am afraid to clutter the table if we don't clean the table from failed attempts. Perhaps we should be a bit more lax on the timing however (a week ?).

Context from the failure should already be reported in `error` property in the response.

Reply:
How is someone supposed to get feedback about the automatic scheduled purge failing? We [don't return them in the admin response](https://github.com/matrix-org/synapse/pull/15488#discussion_r1199393106).